flight/targets/brainre1/re1fpga/fpga_drv.c

```diff
@@ -618,6 +565,90 @@ void PIOS_RE1FPGA_Set3DConfig(enum pios_video_3d_mode mode, uint8_t x_shift_righ
 	PIOS_RE1FPGA_WriteReg(RE1FPGA_REG_XCFG2, cfg, 0xFF);
 }
 
+// **********************************************************************************
+// PIOS WS2811 support functions
+
+struct ws2811_pixel_data_s {
+	uint8_t g;
+	uint8_t r;
+	uint8_t b;
+};
+
+struct ws2811_dev_s {
+#define WS2811_MAGIC 0x31313832		/* '2811' */
+	uint32_t magic;
+	uint16_t max_leds;
+	uint16_t max_idx;
+	struct ws2811_pixel_data_s pixel_data[0];
+};
+
+int PIOS_WS2811_init(ws2811_dev_t *dev_out, const struct pios_ws2811_cfg *cfg,
+		int max_leds)
+{
+	PIOS_Assert(max_leds > 0);
+	PIOS_Assert(max_leds <= 1024);
+
+	ws2811_dev_t dev = PIOS_malloc(sizeof(*dev) +
+			sizeof(struct ws2811_pixel_data_s) * max_leds);
+
+	if (!dev) {
+		return -1;
+	}
+
+	dev->magic = WS2811_MAGIC;
+	dev->max_leds = max_leds;
+	dev-> max_idx = 0;
+
+	PIOS_WS2811_set_all(dev, 0, 0, 0);
+
+	*dev_out = dev;
+
+	return 0;
+}
+
+void PIOS_WS2811_set(ws2811_dev_t ws2811_dev, int idx, uint8_t r, uint8_t g,
+		uint8_t b)
+{
+	PIOS_Assert(ws2811_dev->magic == WS2811_MAGIC);
+
+	if (idx >= ws2811_dev->max_leds) {
+		return;
+	}
+
+	if (idx > ws2811_dev->max_idx) {
+		ws2811_dev->max_idx = idx;
+	}
+
+	ws2811_dev->pixel_data[idx] = (struct ws2811_pixel_data_s)
+			{ .r = r, .g = g, .b = b };
+}
+
+void PIOS_WS2811_set_all(ws2811_dev_t ws2811_dev, uint8_t r, uint8_t g,
+		uint8_t b)
+{
+	PIOS_Assert(ws2811_dev->magic == WS2811_MAGIC);
+
+	for (int i = 0; i < ws2811_dev->max_leds; i++) {
+		PIOS_WS2811_set(ws2811_dev, i, r, g, b);
+	}
+}
+
+void PIOS_WS2811_trigger_update(ws2811_dev_t ws2811_dev)
+{
+	uint16_t n_leds;
+	PIOS_Assert(ws2811_dev->magic == WS2811_MAGIC);
+
+	if (PIOS_RE1FPGA_ClaimBus() != 0)
+		return;
+
+	n_leds = MIN(ws2811_dev->max_idx + 1, 1024);
```

Comment:
Is this a threshold you tested this at?  (Since this is happening in System thread)